File: tests/syquex/test_texto.py
"""
FASE 24 — Test de Texto (Manual 3 §12.1).

TDD: este test ES la especificación. Si _syn_texto_* no existen,
el test C NO compila — eso es correcto. Se corrige el CÓDIGO, no el test.

Manual 3 §12.1: lib/texto.syq — Manipulación avanzada de cadenas
Comando: pytest tests/syquex/test_texto.py -v
Criterio: todas las operaciones de cadena correctas
"""
import os
import subprocess
import sys

import pytest

pytestmark = pytest.mark.syquex

PROJECT_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))

BIN_NAME = "test_texto.exe"
BIN_ABS = os.path.join(PROJECT_ROOT, "tests", BIN_NAME)
TEST_SRC = os.path.join(PROJECT_ROOT, "tests", "test_texto.c")


def _find_gcc():
    gcc_candidate = os.path.join(PROJECT_ROOT, "toolchain_gcc12", "mingw64", "bin", "gcc.exe")
    if os.path.exists(gcc_candidate):
        return gcc_candidate
    for candidate in ("gcc", "gcc.exe"):
        try:
            subprocess.run([candidate, "--version"], capture_output=True)
            return candidate
        except FileNotFoundError:
            continue
    return gcc_candidate


@pytest.fixture(scope="module")
def exe_path():
    """Compila el test C y retorna la ruta al ejecutable."""
    if os.path.exists(BIN_ABS):
        return BIN_ABS

    gcc = _find_gcc()
    r = subprocess.run(
        [gcc, "-O2", "-I", PROJECT_ROOT, "-I.", "-o", BIN_ABS, TEST_SRC,
         "-lm", "-lpthread", "-lws2_32"],
        capture_output=True, text=True, timeout=120
    )
    if r.returncode != 0:
        pytest.fail(
            f"test_texto.c NO COMPILA (TDD: falta implementar §12.1).\n"
            f"Error de gcc:\n{r.stderr[-1000:]}"
        )
    assert os.path.exists(BIN_ABS), f"{BIN_NAME} no se creó"
    return BIN_ABS


def _run(exe_path):
    """Ejecuta el test C y retorna (returncode, stdout, stderr)."""
    r = subprocess.run(
        [exe_path], capture_output=True, cwd=PROJECT_ROOT,
        timeout=30, encoding="utf-8", errors="replace"
    )
    return r.returncode, r.stdout, r.stderr


class TestTextoLongitud:
    """§12.1 — Longitud de cadena."""

    def test_longitud_vacia(self, exe_path):
        rc, out, err = _run(exe_path)
        assert rc == 0, f"test_texto falló:\n{out}\n{err}"
        assert 'longitud("") == 0' in out

    def test_longitud_normal(self, exe_path):
        rc, out, err = _run(exe_path)
        assert 'longitud("hola") == 4' in out


class TestTextoSubcadena:
    """§12.1 — Subcadena."""

    def test_subcadena(self, exe_path):
        rc, out, err = _run(exe_path)
        assert rc == 0, f"test_texto falló:\n{out}\n{err}"
        assert 'subcadena("hola mundo", 0, 4)' in out

    def test_subcadena_vacia(self, exe_path):
        rc, out, err = _run(exe_path)
        assert 'subcadena("abc", 1, 1)' in out


class TestTextoContiene:
    """§12.1 — Contiene."""

    def test_contiene_true(self, exe_path):
        rc, out, err = _run(exe_path)
        assert rc == 0, f"test_texto falló:\n{out}\n{err}"
        assert 'contiene("hola mundo", "mundo") == true' in out

    def test_contiene_false(self, exe_path):
        rc, out, err = _run(exe_path)
        assert 'contiene("hola mundo", "xyz") == false' in out


class TestTextoReemplazar:
    """§12.1 — Reemplazar."""

    def test_reemplazar(self, exe_path):
        rc, out, err = _run(exe_path)
        assert rc == 0, f"test_texto falló:\n{out}\n{err}"
        assert 'reemplazar("hola mundo", "mundo", "syquex")' in out

    def test_reemplazar_multiples(self, exe_path):
        rc, out, err = _run(exe_path)
        assert 'reemplazar("aaa", "a", "b") == "bbb"' in out


class TestTextoDividir:
    """§12.1 — Dividir (split)."""

    def test_dividir(self, exe_path):
        rc, out, err = _run(exe_path)
        assert rc == 0, f"test_texto falló:\n{out}\n{err}"
        assert 'dividir("a,b,c", ",") tiene 3 partes' in out


class TestTextoUnir:
    """§12.1 — Unir (join)."""

    def test_unir(self, exe_path):
        rc, out, err = _run(exe_path)
        assert rc == 0, f"test_texto falló:\n{out}\n{err}"
        assert 'unir([a,b,c], " + ") == "a + b + c"' in out


class TestTextoRecortar:
    """§12.1 — Recortar (trim)."""

    def test_recortar(self, exe_path):
        rc, out, err = _run(exe_path)
        assert rc == 0, f"test_texto falló:\n{out}\n{err}"
        assert 'recortar("  hola  ") == "hola"' in out


class TestTextoMayusculas:
    """§12.1 — Mayúsculas / Minúsculas."""

    def test_mayusculas(self, exe_path):
        rc, out, err = _run(exe_path)
        assert rc == 0, f"test_texto falló:\n{out}\n{err}"
        assert 'mayusculas("hola") == "HOLA"' in out

    def test_minusculas(self, exe_path):
        rc, out, err = _run(exe_path)
        assert 'minusculas("HOLA") == "hola"' in out


class TestTextoOtros:
    """§12.1 — Comienza con, termina con, indice de, repetir, invertir."""

    def test_comienza_con(self, exe_path):
        rc, out, err = _run(exe_path)
        assert rc == 0, f"test_texto falló:\n{out}\n{err}"
        assert 'comienza_con("hola mundo", "hola") == true' in out

    def test_termina_con(self, exe_path):
        rc, out, err = _run(exe_path)
        assert 'termina_con("hola mundo", "mundo") == true' in out

    def test_indice_de(self, exe_path):
        rc, out, err = _run(exe_path)
        assert 'indice_de("hola mundo", "mundo") == 5' in out

    def test_repetir(self, exe_path):
        rc, out, err = _run(exe_path)
        assert 'repetir("ab", 3) == "ababab"' in out

    def test_invertir(self, exe_path):
        rc, out, err = _run(exe_path)
        assert 'invertir("abc") == "cba"' in out


class TestTextoIntegracion:
    """Tests de integración completos del módulo texto."""

    def test_todos_los_tests_c_pasan(self, exe_path):
        rc, out, err = _run(exe_path)
        assert rc == 0, f"test_texto falló (rc={rc}):\n{out}\n{err}"
        assert "RESULTADO" in out, f"Sin resultados en stdout:\n{out}"
        assert "[FAIL]" not in out, f"Hay FAILs en output C:\n{out}"

File: tests/test_texto.c
// FASE 24 — Test de Texto (Manual 3 §12.1)
// TDD: este test ES la especificación. Si las funciones _syn_texto_*
// no existen, el test NO compila — eso es correcto.
//
// Manual 3 §12.1: lib/texto.syq — Manipulación avanzada de cadenas
// Comando: pytest tests/syquex/test_texto.py -v
// Criterio: todas las operaciones de cadena correctas

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include "synapse_rt_types.h"
#include "runtime/core/texto.h"

static int passed = 0;
static int failed = 0;

#define CHECK(cond, msg) do { \
    if (!(cond)) { printf("  [FAIL] %s\n", msg); failed++; } \
    else { printf("  [PASS] %s\n", msg); passed++; } \
} while(0)

#define CS(s) ((CadenaSegura){ .longitud = (int)strlen(s), .datos = (s) })

int main(void) {
    setbuf(stdout, NULL);

    // === 1. Longitud ===
    printf("=== 1. Longitud ===\n");
    CHECK(_syn_texto_longitud(CS("")) == 0, "longitud(\"\") == 0");
    CHECK(_syn_texto_longitud(CS("hola")) == 4, "longitud(\"hola\") == 4");
    CHECK(_syn_texto_longitud(CS("hello world")) == 11, "longitud(\"hello world\") == 11");
    CHECK(_syn_texto_longitud((CadenaSegura){0, NULL}) == 0, "longitud(NULL) == 0");

    // === 2. Subcadena ===
    printf("=== 2. Subcadena ===\n");
    CadenaSegura s2 = _syn_texto_subcadena(CS("hola mundo"), 0, 4);
    CHECK(s2.longitud == 4 && memcmp(s2.datos, "hola", 4) == 0, "subcadena(\"hola mundo\", 0, 4) == \"hola\"");
    CadenaSegura s2b = _syn_texto_subcadena(CS("hola mundo"), 5, 10);
    CHECK(s2b.longitud == 5 && memcmp(s2b.datos, "mundo", 5) == 0, "subcadena(\"hola mundo\", 5, 10) == \"mundo\"");
    CadenaSegura s2c = _syn_texto_subcadena(CS("abc"), 1, 1);
    CHECK(s2c.longitud == 0, "subcadena(\"abc\", 1, 1) == \"\"");

    // === 3. Contiene ===
    printf("=== 3. Contiene ===\n");
    CHECK(_syn_texto_contiene(CS("hola mundo"), CS("mundo")) == 1, "contiene(\"hola mundo\", \"mundo\") == true");
    CHECK(_syn_texto_contiene(CS("hola mundo"), CS("xyz")) == 0, "contiene(\"hola mundo\", \"xyz\") == false");
    CHECK(_syn_texto_contiene(CS("hola"), CS("")) == 1, "contiene(\"hola\", \"\") == true");
    CHECK(_syn_texto_contiene(CS(""), CS("a")) == 0, "contiene(\"\", \"a\") == false");

    // === 4. Reemplazar ===
    printf("=== 4. Reemplazar ===\n");
    CadenaSegura r4 = _syn_texto_reemplazar(CS("hola mundo"), CS("mundo"), CS("syquex"));
    CHECK(r4.longitud == 11 && memcmp(r4.datos, "hola syquex", 11) == 0, "reemplazar(\"hola mundo\", \"mundo\", \"syquex\")");
    CadenaSegura r4b = _syn_texto_reemplazar(CS("aaa"), CS("a"), CS("b"));
    CHECK(r4b.longitud == 3 && memcmp(r4b.datos, "bbb", 3) == 0, "reemplazar(\"aaa\", \"a\", \"b\") == \"bbb\"");
    CadenaSegura r4c = _syn_texto_reemplazar(CS("abc"), CS("x"), CS("y"));
    CHECK(r4c.longitud == 3 && memcmp(r4c.datos, "abc", 3) == 0, "reemplazar sin match retorna original");

    // === 5. Dividir (split) ===
    printf("=== 5. Dividir ===\n");
    int64_t sp = _syn_texto_dividir(CS("a,b,c"), CS(","));
    CHECK(_syn_texto_dividir_longitud(sp) == 3, "dividir(\"a,b,c\", \",\") tiene 3 partes");
    CadenaSegura sp0 = _syn_texto_dividir_obtener(sp, 0);
    CHECK(sp0.longitud == 1 && sp0.datos[0] == 'a', "parte[0] == \"a\"");
    CadenaSegura sp2 = _syn_texto_dividir_obtener(sp, 2);
    CHECK(sp2.longitud == 1 && sp2.datos[0] == 'c', "parte[2] == \"c\"");
    _syn_texto_dividir_liberar(sp);

    // === 6. Unir (join) ===
    printf("=== 6. Unir ===\n");
    int64_t jn = _syn_texto_dividir(CS("a,b,c"), CS(","));
    CadenaSegura j = _syn_texto_unir(jn, CS(" + "));
    CHECK(j.longitud == 9 && memcmp(j.datos, "a + b + c", 9) == 0, "unir([a,b,c], \" + \") == \"a + b + c\"");
    _syn_texto_dividir_liberar(jn);

    // === 7. Recortar (trim) ===
    printf("=== 7. Recortar ===\n");
    CadenaSegura tr = _syn_texto_recortar(CS("  hola  "));
    CHECK(tr.longitud == 4 && memcmp(tr.datos, "hola", 4) == 0, "recortar(\"  hola  \") == \"hola\"");
    CadenaSegura tr2 = _syn_texto_recortar(CS("sin_espacios"));
    CHECK(tr2.longitud == 12, "recortar(\"sin_espacios\") sin cambio");

    // === 8. Mayúsculas / Minúsculas ===
    printf("=== 8. Mayusculas / Minusculas ===\n");
    CadenaSegura mu = _syn_texto_mayusculas(CS("hola"));
    CHECK(mu.longitud == 4 && memcmp(mu.datos, "HOLA", 4) == 0, "mayusculas(\"hola\") == \"HOLA\"");
    CadenaSegura mi = _syn_texto_minusculas(CS("HOLA"));
    CHECK(mi.longitud == 4 && memcmp(mi.datos, "hola", 4) == 0, "minusculas(\"HOLA\") == \"hola\"");

    // === 9. Comienza con / Termina con ===
    printf("=== 9. Comienza con / Termina con ===\n");
    CHECK(_syn_texto_comienza_con(CS("hola mundo"), CS("hola")) == 1, "comienza_con(\"hola mundo\", \"hola\") == true");
    CHECK(_syn_texto_comienza_con(CS("hola mundo"), CS("mundo")) == 0, "comienza_con(\"hola mundo\", \"mundo\") == false");
    CHECK(_syn_texto_termina_con(CS("hola mundo"), CS("mundo")) == 1, "termina_con(\"hola mundo\", \"mundo\") == true");
    CHECK(_syn_texto_termina_con(CS("hola mundo"), CS("hola")) == 0, "termina_con(\"hola mundo\", \"hola\") == false");

    // === 10. Índice de ===
    printf("=== 10. Indice de ===\n");
    CHECK(_syn_texto_indice_de(CS("hola mundo"), CS("mundo")) == 5, "indice_de(\"hola mundo\", \"mundo\") == 5");
    CHECK(_syn_texto_indice_de(CS("hola mundo"), CS("xyz")) == -1, "indice_de(\"hola mundo\", \"xyz\") == -1");
    CHECK(_syn_texto_indice_de(CS("abc"), CS("")) == 0, "indice_de(\"abc\", \"\") == 0");

    // === 11. Repetir ===
    printf("=== 11. Repetir ===\n");
    CadenaSegura rep = _syn_texto_repetir(CS("ab"), 3);
    CHECK(rep.longitud == 6 && memcmp(rep.datos, "ababab", 6) == 0, "repetir(\"ab\", 3) == \"ababab\"");
    CadenaSegura rep0 = _syn_texto_repetir(CS("x"), 0);
    CHECK(rep0.longitud == 0, "repetir(\"x\", 0) == \"\"");

    // === 12. Invertir ===
    printf("=== 12. Invertir ===\n");
    CadenaSegura inv = _syn_texto_invertir(CS("abc"));
    CHECK(inv.longitud == 3 && memcmp(inv.datos, "cba", 3) == 0, "invertir(\"abc\") == \"cba\"");
    CadenaSegura inv1 = _syn_texto_invertir(CS("a"));
    CHECK(inv1.longitud == 1 && inv1.datos[0] == 'a', "invertir(\"a\") == \"a\"");

    printf("\n=== RESULTADO: %d passed, %d failed ===\n", passed, failed);
    return failed > 0 ? 1 : 0;
}
